Optimize WaitForTransaction

ice-coldbell opened the pull request:
### Description
- Optimize WaitForTransaction.

Adopt the WaitForTransaction polling method implemented via [wait_by_hash api](https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/transactions/GET/transactions/wait_by_hash/{txn_hash}) in the Typescript SDK.

However, there is a concern that this might conflict with the PR using Golang context for HTTP requests. 
In that case, a ctx should be accepted and used for polling.

### Test Plan
- Verify that WaitTransactionByHash operates correctly.
- Confirm that PollForTransaction produces the same results as before.

### Related Links
-  https://aptos.dev/en/build/apis/fullnode-rest-api-reference#tag/transactions/GET/transactions/wait_by_hash/{txn_hash}
-  https://github.com/aptos-labs/aptos-go-sdk/pull/128

ice-coldbell commented:
### Question
I'm not entirely sure how the server-side behavior of wait_by_hash works. 
Therefore, I'm wondering whether the PollForTransaction implementation should call TransactionByHash first and then WaitTransactionByHash, or if it's acceptable to skip TransactionByHash and directly call the WaitTransactionByHash API.

gregnazario commented:
> ### Question
> I'm not entirely sure how the server-side behavior of wait_by_hash works. Therefore, I'm wondering whether the PollForTransaction implementation should call TransactionByHash first and then WaitTransactionByHash, or if it's acceptable to skip TransactionByHash and directly call the WaitTransactionByHash API.

The TS SDK has this implementation.

You'll want to call "wait_by_hash" first, and follow it up with the shorter calls.

The Tl;dr is that wait by hash will keep a connection open for that time, and really is not necessary after the first one.  It's better from a long term to call the TransactionByHash call.

ice-coldbell commented:
Fixed the test-related issues. 
Since it's currently not possible to distinguish between an API Error and an HTTP Error, I implemented it in the same way as our existing error handling method. 

Comparing error message strings is not a recommended approach for error handling, so I believe that once an API Error Type is implemented in the future, can add additional handling.

ice-coldbell commented:
thx, greg
